File: src/main/java/eotree/EOObject.java
package eotree;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class EOObject extends EOExpr {
    public List<EOBndName>     freeAttrs;
    public Optional<EOBndName> varargAttr;
    public List<EOBndExpr>     bndAttrs;

    public EOObject(List<EOBndName> freeAttrs,
                    Optional<EOBndName> varargAttr,
                    List<EOBndExpr> bndAttrs) {
        this.freeAttrs = freeAttrs;
        this.varargAttr = varargAttr;
        this.bndAttrs = bndAttrs;
    }

    @Override
    public String generateEO(int indent) {
        return indent(indent) + "[" +
               freeAttrs.stream()
                       .map(attr -> attr.generateEO(indent))
                       .collect(Collectors.joining(" ")) +
               (varargAttr
                       .map(attr -> " " + attr.generateEO(indent) + "...")
                       .orElse("")) +
               "]\n" +
               bndAttrs.stream()
                       .map(attr -> attr.generateEO(indent + 1))
                       .collect(Collectors.joining("\n"));
    }
}

File: src/main/java/eotree/EOObject.kt
package eotree

import arrow.core.Option
import arrow.core.getOrElse
import java.util.*
import java.util.stream.Collectors

class EOObject(
    var freeAttrs: List<String>,
    var varargAttr: Option<String>,
    var bndAttrs: List<EOBndExpr>,
    var comment: String = "",
) : EOExpr() {
    override fun generateEO(indent: Int): String {
        return comment.let {
            if (it.isNotEmpty())
                indent(indent) + it
            else
                it
        } +
                indent(indent) +
                "[" +
                freeAttrs.joinToString(" ") +
                varargAttr
                    .map { attr -> (if (freeAttrs.isNotEmpty()) " " else "") + attr + "..." }
                    .getOrElse { "" } +
                "]" +
                bndAttrs
                    .map { attr: EOBndExpr -> "\n" + attr.generateEO(indent + 1) }
                    .joinToString("")
    }
}